### ISSUE TITLE
Compilation issue with GCC 6.1.1

### ISSUE THREAD

**Issue by ebioman** [PacificBiosciences/pbbam#14]
We migrated from CentOS 6 with gcc4.9 to CentOS 7 with gcc 6.1.1.
While trying to rebuild pbbam for that architecture I ran into issues with gtest.
It looks as if make is somehow trying to change the Dprivate and Dpublic flags and the error results in a ``redeclared with different access`` message.

```c
[ 68%] Building CXX object tests/CMakeFiles/test_pbbam.dir/src/test_AlignmentPrinter.cpp.o
In file included from /builddir/build/BUILD/gtest/include/gtest/internal/gtest-port.h:212:0,
                 from /builddir/build/BUILD/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /builddir/build/BUILD/gtest/include/gtest/gtest.h:58,
                 from /builddir/build/BUILD/pbbam-0106cdf0a2c71ed3195d971b6972b6200dfe66f3/tests/src/test_Accuracy.cpp:42:
/software/include/c++/6.1.1/sstream:300:7: error: 'struct std::__cxx11::basic_stringbuf<_CharT, _Traits, _Alloc>::__xfer_bufptrs' redeclared with different access
       struct __xfer_bufptrs
       ^~~~~~
In file included from /builddir/build/BUILD/gtest/include/gtest/internal/gtest-port.h:212:0,
                 from /builddir/build/BUILD/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /builddir/build/BUILD/gtest/include/gtest/gtest.h:58,
                 from /builddir/build/BUILD/pbbam-0106cdf0a2c71ed3195d971b6972b6200dfe66f3/tests/src/test_AlignmentPrinter.cpp:43:
/software/include/c++/6.1.1/sstream:300:7: error: 'struct std::__cxx11::basic_stringbuf<_CharT, _Traits, _Alloc>::__xfer_bufptrs' redeclared with different access
       struct __xfer_bufptrs
       ^~~~~~
[ 69%] Building CXX object tools/pbindex/CMakeFiles/pbindex.dir/src/main.cpp.o
make[2]: *** [tests/CMakeFiles/test_pbbam.dir/src/test_Accuracy.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

**Comment by ebioman** [comment]
I tried to change the definitions for your gtests

by replacing ``#define private public`` with ``#define private private`` which resulted in other errors due to access problems.
Subsequently I declared in the ``CMakeLists.txt`` PacBioBAM's unit tests as off.
This works but obviously is only a work-around instead of a fix..

<pre><code>
--- CMakeLists.txt.o    2017-06-08 09:58:59.889735232 +0200
+++ CMakeLists.txt      2017-06-08 09:59:50.523680131 +0200
@@ -14,7 +14,7 @@

 # list build-time options
 option(PacBioBAM_build_docs    "Build PacBioBAM's API documentation."                   ON)
-option(PacBioBAM_build_tests   "Build PacBioBAM's unit tests."                          ON)
+option(PacBioBAM_build_tests   "Build PacBioBAM's unit tests."                          OFF)
 option(PacBioBAM_build_shared  "Build PacBioBAM as shared library as well."             OFF)
 option(PacBioBAM_build_tools   "Build PacBioBAM command line utilities (e.g. pbindex)"  ON)
 option(PacBioBAM_wrap_csharp   "Build PacBioBAM with SWIG bindings for C#."             OFF)
</code></pre>

**Comment by pb-cdunn** [comment]
Yes, this is a serious problem for the pbbam tests, as currently written.
```
$ git grep 'define private public' | wc -l
35
```

For VS2012+, we use this trick in `tests/CMakeLists.txt`:
```
if(MSVC)
    # VS2012+ pooh-pooh's Derek's "#define private public" trick
    add_definitions(-D_ALLOW_KEYWORD_MACROS)
endif()
```
Ultimately, we need some refactoring to make the internals testable. Short term, the simplest solution is to make everything explicitly public.

**Comment by jwakely** [comment]
This is batshit crazy. GCC has a `-fno-access-control` flag which might be a better way to do whatever it is you're trying to do, without redefining C++ keywords.

**Comment by jwakely** [comment]
Alternatively, only redefine `private` after including all third-party headers, so it only affects your own code and not standard headers like `<sstream>`. That's still disgusting and crazy, but it might just work.

**Comment by pezmaster31** [comment]
No, I completely agree. That was purely an admittedly gross hack for testing, from back in the very initial development phases. It just hadn't been removed yet. That said, I'm well aware and it's already on my board to remove it soon. 

Thanks for the feedback and apologies about any convenience in the meantime.
 - Derek

**Comment by SoapZA** [comment]
I have shifted the headers around and the definition of the macro hack, pbbam builds fine for us with GCC 6 and 7 for some time already.
@jwakely Agreed, redefining C++ keywords is iffy, unfortunately there'y no portable `-fno-access-control` across compilers. Once we got rid of our CMake stuff, we might just switch to using `-fno-access-control` from Meson, but for now, correct placement of `#define private public` allows us to build fine.
